### DISCUSSION TITLE
Split difficulty adjust into CS, AR, OD, and HP adjusts

### DISCUSSION THREAD

**Discussion by Natelytle** [ppy/osu#24961] [Ideas]
A small discussion was held in the PP server about the feasibility of displaying the difficulty adjust attributes similarly to how DT displays the rate, and a solution that was presented that is quite elegant in my opinion was to split the mod into 4 smaller scope mods, and display the difficulty value of each like how DT rate is displayed. This isn't the only reason either, there's a few more I can think of for this being a change that may be a good way forward:
- Splitting the mod out into 4 separate mods would make score multipliers easier to balance. A simple approach would be to set each mod to be a 0.85x multiplier, so having all 4 mods would be a similar multiplier to current DA, but you get punished less for only changing a single attribute. This would also make balancing custom multipliers for the difficulty stats easier and more intuitive, as you don't have to think of the multipliers in terms of each other.
- It would make it more intuitive and easier to parse what someone changed if they have a smaller scope mod, vs the all encompassing difficulty adjust. If someone has a single difficulty adjustment mod (for example, CS), you can tell that at a glance vs needing to hover over the score. That may not be a big problem if you're checking out individual scores, but parsing a map's leaderboard as a whole becomes a pain with the current difficulty adjust mod.
- Difficulty adjust including all of the settings together brings the connotation of it being a min-max mod, which doesn't sit well with certain players. Splitting them could make it easier to stick to changing individual settings to retain some score multiplier, which also helps retain the mapper's intention with the difficulty settings.
- This would also make it possible for multis and playlists to allow you to change certain difficulty settings but not others. Right now, setting difficulty adjust as allowed allows all of the difficulty settings to be changed, and there's no good UX way to change this while retaining it being a single mod,

However, there are a few drawbacks to this. It contributes a little to the "mod clutter" that some people aren't fond of, and also makes adjusting the mods a bit annoying when you have all 4 selected because they'll be lined up horizontally and not vertically. If conversion of old scores is something you'd want as well, I'm not sure of the complexity of that either. Personally, I think these drawbacks aren't as bad as the benefits of this change, but in the end its up to what is seen as more intuitive.

**Comment by peppy** [comment]
I'm not against this, but also not 100% sure. "Difficulty adjust" has a nice ring to it and you go in knowing what to expect. It also allows a catch-all mod which can be overloaded with future difficulty settings.

But in terms of leaderboards, I definitely agree it's a painful one to display, process and compare scores.

In terms of adding new mods names after the difficulty attributes, all acronyms are free except for "CS" (used by "Constant Speed"). We would probably want to change "CS" to something else while we can, to avoid any confusion.

Will wait for more opinions.

**Reply by Rekunan** [reply]
What about "SZ" for size? Since in mania, CS is the key count, so we could also merge the keyconvert mod with this.

**Reply by peppy** [reply]
There's no room for discussion here. If we're going with this, then "Constant speed" will be renamed.

**Reply by Givikap120** [reply]
I think CS can be exclusive to rulesets where it makes sense. Same with other difficulty attributes.
Custom rulesets can implement their own diff attribute and a mod to change it with Difficulty Adjust mod being an abstract parent class.

**Comment by apollo-dw** [comment]
Having it all in one mod is super convenient, especially keeping in mind just how many mods we currently have as-is. I also don't think these benefits described aren't as huge as they're described in the OP. To give some more detail:

> Splitting the mod out into 4 separate mods would make score multipliers easier to balance. 

This is untrue - the difficulty of balancing wouldn't change at all. The only difference here would be that they'd be split into four different files instead of being together (not trying to imply one way is better than the other). You can follow the philosophy of _"you don't have to think of the multipliers in terms of each other"_ whether or not they are in the same file, it's all dependent on implementation.

> It would make it more intuitive and easier to parse what someone changed if they have a smaller scope mod, vs the all encompassing difficulty adjust.

I don't see this as being mutually exclusive with the difficulty adjust mod - as I said we can consider alternative means of displaying changed settings without changing the mod. Nothing impossible there. If you were thinking of using the new display used for rate changes, then consider how four difficulty settings + DT will look like. 

> Difficulty adjust including all of the settings together brings the connotation of it being a min-max mod

... I just disagree with this? Outside of this seeming like a nothing burger, min-maxxers will still min-max regardless of if it's in one mod or four or eight. 

> This would also make it possible for multis and playlists to allow you to change certain difficulty settings but not others. 

I can see how splitting these mods up could make it simpler for the user, but there's still ways around this with the "Difficulty Adjust" mod. It's very likely that eventually there will *need* to be host-placed freemod restrictions on mod settings anyways because of [every single other mod that exists], so this isn't going to be an unsolved problem forever. Naive proposal from me would be dual-range slider and enable/disable toggle per setting.

**Comment by LeNitrous** [comment]
A lot of internals regarding PP calculation and mods are wholly separate from one another so splitting up "difficulty adjust" makes no sense. If the concern is presentation towards the end user, it is preferrable to make better UX at conveying the information.

A lot of customizable mods have one or more setting, should that mean, we should also split those up for the sake of presentation? Currently (as of this writing) customized mods on leaderboards show only 1 customized setting (usually the most prominent one) while having more than one setting.

What needs to happen is changing the UX. For mods with multiple configurable settings, an indicator needs to be shown that there are customizations applied to the mod. In my opinion, this is more than enough information to convey at a leaderboard display. However, at a macro-level, the leaderboards and results screen/score display may need to be redesigned to better convey this information.

**Reply by peppy** [reply]
I don't quite understand where you're coming from. Could you better explain the "pp internals" part?

You said a lot of words but I find what you are implying is detrimental from a change like this.

**Comment by LeNitrous** [comment]
> I don't quite understand where you're coming from. Could you better explain the "pp internals" part?
> 
> You said a lot of words but I find what you are implying is detrimental from a change like this.

Basically, score calculation is a separate system from mods. What I'm proposing is basically a better way to present adjusted/customized mods through UX without changing any of the existing mods.

**Reply by peppy** [reply]
I believe you're missing the point. The "indicator light" is still going to happen, but this issue is ciscussing the ability to show the most important statistics to ease the user experience. No user wants to hover 50 mods to see what rate adjust people have applied.

**Comment by Stoppedpuma** [comment]
I think splitting up difficulty adjust is the wrong way to go as it adds a lot of unnecessary clutter and makes it less convenient to change multiple values.

> Splitting the mod out into 4 separate mods would make score multipliers easier to balance.

This doesn't make sense to me, score multipliers could vary based on the value type changed in difficulty adjust as they do in DT and HT  (e.g. 0.02x per +1 CS / 0.01x per +1 AR) by using the sliders to change value. Why change it if splitting it out to 4 mods is going to do the exact same thing but just add more mods to the menu?

> This would also make it possible for multis and playlists to allow you to change certain difficulty settings but not others.

I agree with being able to allow change to only certain values, putting a lock icon next to the value in mod customization could work for keeping it a single mod instead of 4.

Design wise it also looks better to my eyes to keep it as one mod and just show changed values instead of splitting it into 4 mods:
![osu!_song_select_diffadj_split](https://github.com/ppy/osu/assets/58333920/c773884f-6945-4714-97ac-2d88ab4cd03b)

**Reply by Givikap120** [reply]
adding score for increasing AR is stupid
here is my idea for coefs

CS DA: +0.02 multiplier for every 1 CS increase, -0.08 multiplier for every 1 CS decrease
HP DA: unchanged on increasing HP, -0.05 multiplier for every 1 HP decrease
AR DA: -0.03 multiplier decrease for every 0.1 AR change in any direction, capped at 0.7 multiplier
OD DA: +0.01 multiplier for every 1 OD increase, -0.05 multiplier for every 1 OD decrease

**Reply by Stoppedpuma** [reply]
I was using the values as an example, a discussion (probably separate from here to not get off topic) is definitely needed for difficulty adjust multipliers though.

**Comment by Natelytle** [comment]
Now that I'm using difficulty adjust a lot more recently, one giant point of contention with the current system is that the difficulty adjust mod leaderboards are just useless on their own. 95% of difficulty adjust usages are just to change the AR, and as someone who uses it to play CS8 on everything it will be really difficult to find other people using it to play higher CS values on maps when the leaderboards get filled.
